### PR TITLE
Don't use exchange in the hot path of the GC

### PR DESCRIPTION
**Pull request by gbaraldi**
`xchg` is a very expensive instruction when working with memory, on x86 (intel 11th gen) it has an implied lock and a latency of 19. A false positive here just means a slight amount of extra work so don't do that.

### CODE REVIEW

File: src/gc.c [hunk 1]
```diff
@@ -818,9 +818,10 @@ FORCE_INLINE int gc_try_setmark_tag(jl_taggedvalue_t *o, uint8_t mark_mode) JL_N
         tag = tag | mark_mode;
         assert((tag & 0x3) == mark_mode);
     }
-    tag = jl_atomic_exchange_relaxed((_Atomic(uintptr_t)*)&o->header, tag);
+    uintptr_t curr = jl_atomic_load_relaxed((_Atomic(uintptr_t)*)&o->header);
```
Comment (vchuravy) [inline src/gc.c:821]: Should we make the load in 807 also relaxed to match here?

Also as we discussed on slack not sure that the load here is super necessary (can you disassemble this function?) / Wouldn't need a compiler barrier.

File: src/gc.c [hunk 2]
```diff
@@ -818,9 +818,10 @@ FORCE_INLINE int gc_try_setmark_tag(jl_taggedvalue_t *o, uint8_t mark_mode) JL_N
...
+    jl_atomic_store_relaxed((_Atomic(uintptr_t)*)&o->header, tag);
```
Comment (vchuravy) [inline src/gc.c:822]: Can you add a comment as well so that some poor soul doesn't change this back to a xchg

File: src/gc.c (hunk 1 above)
Comment (gbaraldi) [reply src/gc.c:821]: By this you mean making the load above atomic_relaxed and just reusing it? That's fine by me.

File: src/gc.c [hunk 3]
```diff
@@ -818,9 +819,8 @@ FORCE_INLINE int gc_try_setmark_tag(jl_taggedvalue_t *o, uint8_t mark_mode) JL_N
         tag = tag | mark_mode;
         assert((tag & 0x3) == mark_mode);
     }
-    uintptr_t curr = jl_atomic_load_relaxed((_Atomic(uintptr_t)*)&o->header);
-    jl_atomic_store_relaxed((_Atomic(uintptr_t)*)&o->header, tag);
-    verify_val(jl_valueof(o));
+    jl_atomic_store_relaxed((_Atomic(uintptr_t)*)&o->header, tag); //xchg here was slower than
+    verify_val(jl_valueof(o));                                     //potentially redoing work because of a stale tag.
     return !gc_marked(curr);
```
Comment (vchuravy) [inline src/gc.c:824]: Let's return a constant here now,

**Review by vchuravy — APPROVED** [review]

**Comment by oscardssmith** [comment]
does this make a noticeable difference in benchmarks?

**Comment by vchuravy** [comment]
@d-netto I think this is correct and I think @gbaraldi had a benchmark example where this saved a few couple of percent.

File: src/gc.c [hunk 4]
```diff
@@ -818,9 +818,9 @@ FORCE_INLINE int gc_try_setmark_tag(jl_taggedvalue_t *o, uint8_t mark_mode) JL_N
         tag = tag | mark_mode;
         assert((tag & 0x3) == mark_mode);
     }
-    tag = jl_atomic_exchange_relaxed((_Atomic(uintptr_t)*)&o->header, tag);
-    verify_val(jl_valueof(o));
-    return !gc_marked(tag);
+    jl_atomic_store_relaxed((_Atomic(uintptr_t)*)&o->header, tag); //xchg here was slower than
+    verify_val(jl_valueof(o));                                     //potentially redoing work because of a stale tag.
+    return 1;
```
Comment (d-netto) [inline src/gc.c:823]: I think this should still return `!gc_marked(tag)`.

Comment (d-netto) [reply src/gc.c:823]: Ah, OK. Just saw the:

```
if (gc_marked(tag))
    return 0;
```

**Review by d-netto — APPROVED** [review]
Seems mergeable to me.

**Pull request merged**